### ISSUE TITLE
RTC::Consumer::GetStats() | failed assertion `it != this->mapProfileRtpStream.end()': effective profile does not map to a rtp stream

### ISSUE THREAD

**Issue by ibc** [versatica/mediasoup#160]
```
INFO:Participant#db6bivay consumer "effectiveprofilechange" event [profile:high]
INFO:Participant#db6bivay consumer "effectiveprofilechange" event [profile:none]
INFO:Participant#db6bivay consumer "effectiveprofilechange" event [profile:medium]
INFO:Participant#jxull4dn consumer "effectiveprofilechange" event [profile:high]
mediasoup-worker's stderr: ABORT[id:zxvppdey#1] RTC::Consumer::GetStats() | failed assertion `it != this->mapProfileRtpStream.end()': effective profile does not map to a rtp stream
mediasoup:ERROR:Worker child process exited [id:zxvppdey#1, code:null, signal:SIGABRT]
```

It happened once having stats retrieval enabled.

**Comment by ibc** [comment]
```c++
			if (this->effectiveProfile != RTC::RtpEncodingParameters::Profile::NONE)
			{
				auto it = this->mapProfileRtpStream.find(this->effectiveProfile);

				MS_ASSERT(
				  it != this->mapProfileRtpStream.end(), "effective profile does not map to a rtp stream");
```

That is wrong. Effective profile may change at any time, or all the profiles may dissapear, etc.

**Comment by ibc** [comment]
It seems that the same may crash on `Consumer::GetLossPercentage()`.

**Comment by ibc** [comment]
It seems to me that `effectiveProfile` sometimes may point to `targetProfile` while such a profile does not exist in the map.

**Comment by jmillan** [comment]
> It seems to me that effectiveProfile sometimes may point to targetProfile while such a profile does not exist in the map.

your suspicion is correct :-)

**Comment by ibc** [comment]
It has crashed again in 2.0.4:

```
mediasoup-worker's stderr: ABORT[id:yoydtkfb#1] RTC::Consumer::GetStats() | failed assertion `it != this->mapProfileRtpStream.end()': effective profile does not map to a rtp stream
```

**Comment by ibc** [comment]
It happens very often by just running:

```js
CONSUMER.on('stats', (stats) => console.warn(JSON.stringify(stats, null, '  ')));
CONSUMER.enableStats();
```

AFAIS mostly when high stream dies.

**Comment by ibc** [comment]
Definitely we are not setting `this->targetProfile` at all in this case:

https://github.com/versatica/mediasoup/blob/master/worker/src/RTC/Consumer.cpp#L1067

so `targetProfile` may point to a deleted profile/stream.

**Comment by ibc** [comment]
Once `RecalculateTargetProfile()` decides which the `newTargetProfile` is, it should set it as `this->targetProfile` but, instead of that, it just sets it in some of the `if / else` below the assignment. IMHO it should do it at the beginning.

**Comment by ibc** [comment]
Some problem here: https://github.com/versatica/mediasoup/blob/master/worker/src/RTC/Consumer.cpp#L1047
`this->targetProfile` is not set so it may point to a dead one.

**Comment by ibc** [comment]
Ok, just forgot all I said above. It's `this->effectiveProfile` who is causing the crash.

**Comment by ibc** [comment]
AFAIS this commit e28738f43fab720b60d99430423e8769fa366993 fixed nothing. `RecalculateTargetProfile()` does not call to `SetEffectiveProfile()` at all.

**Comment by ibc** [comment]
Proposal: we need a test unit for the `Consumer` that provides source profiles, remove them, and checks that the effective and target profiles are always the proper ones.

**Comment by ibc** [comment]
a961e11e5f857aa59ed51cdab6782f0a7bca73ce seems to fix the crash, and I would also say that I don't see frozen video anymore (#164) but I cannot confirm it at 100%.

Said that, there is something I don't like in the `Consumer::RemoveProfile()` method:

https://github.com/versatica/mediasoup/blob/master/worker/src/RTC/Consumer.cpp#L356

Why are we setting there the new `targetProfile` rather than relying on the `RecalculateTargetProfile()` method? IMHO we should rely on a single method to set the target profile (once we enable the effective one and add/remove available profiles).

**Comment by ibc** [comment]
Theoretically fixed in 2.0.5
